File: lib/audioengine_web/channels/control_channel.ex
defmodule AudioengineWeb.ControlChannel do
  use Phoenix.Channel

  def join("channel:control", _message, socket) do
    {:ok, socket}
  end

  # def join("channel:" <> _private_room_id, _params, _socket) do
  #   {:error, %{reason: "unauthorized"}}
  # end

  def handle_in("update", %{"body" => body}, socket) do
    broadcast! socket, "update", %{body: body}
    {:noreply, socket}
  end
end

File: assets/js/control_channel.js
import { Socket } from "phoenix";

const socket = new Socket("/socket");
socket.connect();
const channel = socket.channel("channel:control", {});

channel.join()
  .receive("ok", resp => { console.log("Joined successfully", resp); })
  .receive("error", resp => { console.log("Unable to join", resp); });

channel.on("update", (payload) => {
  const { body: { key, value } } = payload;
  console.log(payload);
  // const { elem, type } = uiElements[key];

  // switch (type) {
  // case 'slider':
  //   break;
  // case 'toggle':
  //   elem._state.flip(value);
  //   elem.render();
  //   break;
  // case 'button':
  //   break;
  // case 'sequencer':
  //   break;
  // }
});

const pushChange = ({ key, value }) => {
  channel.push("update", { body: { key: key, value: value }});
};

export {
  pushChange
}
